balancer/proposals/balancer.md

```diff
@@ -0,0 +1,180 @@
+
+```go
+// Balancer is an object used to automatically keep the desired number of
+// replicas (pods) distributed among the specified set of targets (deployments
+// or other objects that expose the Scale subresource).
+type Balancer struct {
+   metav1.TypeMeta
+   // Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+   // +optional
+   metav1.ObjectMeta
+   // Specification of the Balancer behavior.
+   // More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+   Spec BalancerSpec
+   // Current information about the Balancer.
+   // +optional
+   Status BalancerStatus
+}
+ 
+// BalancerSpec is the specification of the Balancer behavior.
+type BalancerSpec struct {
+   // Targets is a list of targets between which Balancer tries to distribute
+   // replicas.
+   Targets []BalancerTarget
+   // Replicas is the number of pods that should be distributed among the
+   // declared targets according to the specified policy.
+   Replicas int32
+   // Selector that groups the pods from all targets together (and only those).
+   // Ideally it should match the selector used by the Service built on top of the
+   // Balancer. All pods selectable by targets' selector must match to this selector,
+   // however target's selector don't have to be a superset of this one (although
+   // it is recommended).
+   Selector metav1.LabelSelector
+   // Policy defines how the balancer should distribute replicas among targets.
+   Policy BalancerPolicy
+}
+ 
+// BalancerTarget is the declaration of one of the targets between which the balancer
+// tries to distribute replicas.
+type BalancerTarget struct {
+   // Name of the target.
+   Name string
+   // ScaleTargetRef is a reference that points to a target resource to balance.
+   // The target needs to expose the Scale subresource.
+   ScaleTargetRef hpa.CrossVersionObjectReference
+   // MinReplicas is the minimum number of replicas inside of this target.
+   // Balancer will set at least this amount on the target, even if the total
+   // desired number of replicas for Balancer is lower.
+   // +optional
+   MinReplicas *int32
+   // MaxReplicas is the maximum number of replicas inside of this target.
+   // Balancer will set at most this amount on the target, even if the total
+   // desired number of replicas for the Balancer is higher.
+   // +optional
+   MaxReplicas *int32
+}
+ 
+// BalancerPolicyName is the name of the balancer Policy.
+type BalancerPolicyName string
+const (
+   PriorityPolicyName     BalancerPolicyName = "priority"
+   ProportionalPolicyName BalancerPolicyName = "proportional"
+)
+ 
+// BalancerPolicy defines Balancer policy for replica distribution.
+type BalancerPolicy struct {
+   // PolicyName decides how to balance replicas across the targets.
+   // Depending on the name one of the fields Priorities or Proportions must be set.
+   PolicyName BalancerPolicyName
```

Comment:
We have this pattern in many places, for example when specifying hpa type:  https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/types.go#L217.

Sure, it could work without the name. I don't have strong opinion on that.